pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java

```diff
@@ -915,13 +915,24 @@ public CompletableFuture<Void> connectionOpened(final ClientCnx cnx) {
 
                 resetBackoff();
 
-                boolean firstTimeConnect = subscribeFuture.complete(this);
-                // if the consumer is not partitioned or is re-connected and is partitioned, we send the flow
-                // command to receive messages.
-                if (!(firstTimeConnect && hasParentConsumer) && getCurrentReceiverQueueSize() != 0) {
-                    increaseAvailablePermits(cnx, getCurrentReceiverQueueSize());
-                }
-                future.complete(null);
+                // Switch to user thread to avoid a deadlock as follows:
+                // - create consumer async.
+                // - use IO thread to call a sync method.
+                // - the thread was stuck, and will never respond for brokers.
+                client.externalExecutorProvider().getExecutor(ConsumerImpl.this).execute(() -> {
```

Comment:
The external thread pool is used by message listeners, we usually use internal thread pool for the client's internal tasks.

Reply:
And do we really need to fix this issue? User should not call a sync method in the callback of the async method. This could be something of the scope of Pulsar.

Reply:
Removed the improvement